### ISSUE TITLE
TakeTokens is not thread safe

### ISSUE THREAD

**Issue by HomayoonAlimohammadi** [molBR/bucketgo#1]
Since for each request to the root url the `TakeTokens` function is going to be called, the `b.Tokens` value is going to decrement but not in a concurrency safe way. Try bombarding the server with concurrent requests and you might get lucky and make additional requests. Implementing a lock should solve the issue.

**Comment by molBR** [comment]
I agree, making it thread safe will delay the requests, but it is a good pratice, I'll try it out.
